core/isolate.rs

```diff
@@ -208,18 +228,31 @@ impl Isolate {
     // At this point the SharedQueue should be empty.
     assert_eq!(isolate.shared.size(), 0);
 
-    match op {
-      Op::Sync(buf) => {
-        // For sync messages, we always return the response via Deno.core.send's
-        // return value.
-        // TODO(ry) check that if JSError thrown during respond(), that it will be
-        // picked up.
-        let _ = isolate.respond(Some(&buf));
-      }
-      Op::Async(fut) => {
-        isolate.pending_ops.push(fut);
-        isolate.have_unpolled_ops = true;
-      }
+    match op_result {
+      Ok(op) => match (is_sync, op) {
+        (true, Op::Sync(buf)) => {
+          // For sync messages, we always return the response via Deno.core.send's
+          // return value.
+          // TODO(ry) check that if JSError thrown during respond(), that it will be
+          // picked up.
+          let _ = isolate.respond(Some(&buf));
+        }
+        (false, Op::Async(fut)) => {
+          isolate.pending_ops.push(fut);
+          isolate.have_unpolled_ops = true;
+        }
+        // Panic on unexpected synchronicity. Maybe it would be better to have two
+        // dispatch functions(one for sync one for async).
+        (true, Op::Async(_)) => {
+          panic!("Dispatch returned Op::Async for sync call!")
+        }
+        (false, Op::Sync(_)) => {
+          panic!("Dispatch returned Op::Sync for async call!")
```

Comment:
The sync value here could easily be converted to a future, but should we? I already did this in `cli::ops::dispatch_all_legacy`, but I'm not sure If this should apply to core or not.